Add flex decoding benchmark

joydddd opened the pull request:
ghstack-source-id: b4f26fb66ed47907b11580c8c853737959c58811
Pull Request resolved: https://github.com/pytorch/pytorch/pull/130788

Add benchmark for flex decoding.

joydddd commented:
Flex Decoding performance sweep: 

### Compared to sdpa kernel (w split-K)
bf16 
```
python benchmarks/transformer/score_mod.py --kv-cache-size 32 -s 1024 4096 16384 32768 -d 64 128 -mods noop rel head_bias --decoding --cal-bandwidth -nh 16,16 16,2 16,1
```

| Type    |   Speedup | score_mod     | dtype          | shape(B,Hq,M,Hkv,N,D)     |
|---------|-----------|---------------|----------------|---------------------------|
| Average |     1.121 |               |                |                           |
| Max     |     1.431 | noop          | torch.bfloat16 | (1, 16, 1, 2, 32768, 128) |
| Min     |     0.912 | relative_bias | torch.bfloat16 | (32, 16, 1, 1, 4096, 64)  |


fp32
```
python benchmarks/transformer/score_mod.py --kv-cache-size 32 -s 1024 4096 16384 32768 -d 64 -mods noop rel head_bias --decoding --cal-bandwidth -nh 16,16 16,
2 16,1 -dtype float32
```

| Type    |   Speedup | score_mod   | dtype         | shape(B,Hq,M,Hkv,N,D)    |
|---------|-----------|-------------|---------------|--------------------------|
| Average |    49.452 |             |               |                          |
| Max     |   146.435 | noop        | torch.float32 | (1, 16, 1, 2, 32768, 64) |
| Min     |     5.103 | head_bias   | torch.float32 | (64, 16, 1, 1, 1024, 64) |

<details>
<summary> Full performance sweep </summary>

| score_mod     | dtype          | shape(B,Hq,M,Hkv,N,D)     |   fwd_eager_time |   fwd_compiled_time |   fwd_speedup |   fwd_mem_bw (TB/s) |   GFlops/s |                                      
|---------------|----------------|---------------------------|------------------|---------------------|---------------|---------------------|------------|                                      
| noop          | torch.bfloat16 | (8, 16, 1, 16, 1024, 64)  |           42.539 |              34.481 |         1.234 |               0.886 |    980.735 |                                      
| relative_bias | torch.bfloat16 | (8, 16, 1, 16, 1024, 64)  |           42.554 |              34.741 |         1.225 |               0.879 |    973.395 |                                      
| head_bias     | torch.bfloat16 | (8, 16, 1, 16, 1024, 64)  |           41.852 |              34.341 |         1.219 |               0.890 |    984.729 |                                      
| noop          | torch.bfloat16 | (4, 16, 1, 16, 1024, 128) |           35.436 |              32.709 |         1.083 |               0.934 |   1029.864 |                                      
| relative_bias | torch.bfloat16 | (4, 16, 1, 16, 1024, 128) |           35.418 |              32.910 |         1.076 |               0.928 |   1023.560 |                                      
| head_bias     | torch.bfloat16 | (4, 16, 1, 16, 1024, 128) |           35.391 |              32.789 |         1.079 |               0.932 |   1027.329 |                                      
| noop          | torch.bfloat16 | (2, 16, 1, 16, 4096, 64)  |           37.577 |              40.644 |         0.925 |               0.751 |    832.014 |                                      
| relative_bias | torch.bfloat16 | (2, 16, 1, 16, 4096, 64)  |           37.670 |              40.716 |         0.925 |               0.750 |    830.549 |                                      
| head_bias     | torch.bfloat16 | (2, 16, 1, 16, 4096, 64)  |           37.682 |              40.466 |         0.931 |               0.754 |    835.669 |                                      
| noop          | torch.bfloat16 | (1, 16, 1, 16, 4096, 128) |           39.175 |              35.694 |         1.098 |               0.855 |    943.742 |                                      
| relative_bias | torch.bfloat16 | (1, 16, 1, 16, 4096, 128) |           39.175 |              36.760 |         1.066 |               0.830 |    916.367 |                                      
| head_bias     | torch.bfloat16 | (1, 16, 1, 16, 4096, 128) |           39.192 |              35.700 |         1.098 |               0.855 |    943.563 |                                      
| noop          | torch.bfloat16 | (64, 16, 1, 2, 1024, 64)  |           41.985 |              34.770 |         1.208 |               0.885 |   7780.683 |                                      
| relative_bias | torch.bfloat16 | (64, 16, 1, 2, 1024, 64)  |           41.342 |              35.264 |         1.172 |               0.872 |   7671.566 |                                      
| head_bias     | torch.bfloat16 | (64, 16, 1, 2, 1024, 64)  |           41.576 |              34.726 |         1.197 |               0.886 |   7790.389 |                                      
| noop          | torch.bfloat16 | (32, 16, 1, 2, 1024, 128) |           36.480 |              33.161 |         1.100 |               0.927 |   8126.516 |                                      
| relative_bias | torch.bfloat16 | (32, 16, 1, 2, 1024, 128) |           36.455 |              33.418 |         1.091 |               0.920 |   8064.011 |
| head_bias     | torch.bfloat16 | (32, 16, 1, 2, 1024, 128) |           36.504 |              33.217 |         1.099 |               0.926 |   8112.853 |
| noop          | torch.bfloat16 | (16, 16, 1, 2, 4096, 64)  |           38.040 |              41.179 |         0.924 |               0.743 |   6569.603 |
| relative_bias | torch.bfloat16 | (16, 16, 1, 2, 4096, 64)  |           38.036 |              41.208 |         0.923 |               0.742 |   6565.092 |
| head_bias     | torch.bfloat16 | (16, 16, 1, 2, 4096, 64)  |           38.045 |              41.167 |         0.924 |               0.743 |   6571.547 |
| noop          | torch.bfloat16 | (8, 16, 1, 2, 4096, 128)  |           39.248 |              37.293 |         1.052 |               0.820 |   7226.070 |
| relative_bias | torch.bfloat16 | (8, 16, 1, 2, 4096, 128)  |           39.261 |              37.528 |         1.046 |               0.815 |   7180.942 |
| head_bias     | torch.bfloat16 | (8, 16, 1, 2, 4096, 128)  |           39.235 |              36.226 |         1.083 |               0.844 |   7439.038 |
| noop          | torch.bfloat16 | (4, 16, 1, 2, 16384, 64)  |           41.123 |              38.299 |         1.074 |               0.797 |   7063.629 |
| relative_bias | torch.bfloat16 | (4, 16, 1, 2, 16384, 64)  |           41.144 |              39.322 |         1.046 |               0.776 |   6879.939 |
| head_bias     | torch.bfloat16 | (4, 16, 1, 2, 16384, 64)  |           41.126 |              38.250 |         1.075 |               0.798 |   7072.689 |
| noop          | torch.bfloat16 | (2, 16, 1, 2, 16384, 128) |           44.458 |              37.486 |         1.186 |               0.814 |   7188.891 |
| relative_bias | torch.bfloat16 | (2, 16, 1, 2, 16384, 128) |           44.479 |              36.686 |         1.212 |               0.832 |   7345.666 |
| head_bias     | torch.bfloat16 | (2, 16, 1, 2, 16384, 128) |           44.452 |              37.459 |         1.187 |               0.815 |   7194.196 |
| noop          | torch.bfloat16 | (2, 16, 1, 2, 32768, 64)  |           45.497 |              39.490 |         1.152 |               0.773 |   6850.739 |
| relative_bias | torch.bfloat16 | (2, 16, 1, 2, 32768, 64)  |           45.518 |              39.233 |         1.160 |               0.778 |   6895.620 |
| head_bias     | torch.bfloat16 | (2, 16, 1, 2, 32768, 64)  |           45.524 |              39.440 |         1.154 |               0.774 |   6859.375 |
| noop          | torch.bfloat16 | (1, 16, 1, 2, 32768, 128) |           52.496 |              36.697 |         1.431 |               0.832 |   7343.578 |
| relative_bias | torch.bfloat16 | (1, 16, 1, 2, 32768, 128) |           52.474 |              37.929 |         1.383 |               0.805 |   7104.922 |
| head_bias     | torch.bfloat16 | (1, 16, 1, 2, 32768, 128) |           52.442 |              37.646 |         1.393 |               0.811 |   7158.407 |
| noop          | torch.bfloat16 | (128, 16, 1, 1, 1024, 64) |           41.910 |              34.698 |         1.208 |               0.893 |  15593.514 |
| relative_bias | torch.bfloat16 | (128, 16, 1, 1, 1024, 64) |           41.489 |              35.045 |         1.184 |               0.884 |  15439.191 |
| head_bias     | torch.bfloat16 | (128, 16, 1, 1, 1024, 64) |           41.911 |              34.708 |         1.208 |               0.893 |  15589.070 |
| noop          | torch.bfloat16 | (64, 16, 1, 1, 1024, 128) |           36.202 |              33.375 |         1.085 |               0.929 |  16148.932 |
| relative_bias | torch.bfloat16 | (64, 16, 1, 1, 1024, 128) |           36.200 |              33.514 |         1.080 |               0.925 |  16082.055 |
| head_bias     | torch.bfloat16 | (64, 16, 1, 1, 1024, 128) |           36.200 |              33.381 |         1.084 |               0.929 |  16145.932 |
| noop          | torch.bfloat16 | (32, 16, 1, 1, 4096, 64)  |           37.653 |              41.020 |         0.918 |               0.747 |  13190.352 |
| relative_bias | torch.bfloat16 | (32, 16, 1, 1, 4096, 64)  |           37.682 |              41.317 |         0.912 |               0.742 |  13095.611 |
| head_bias     | torch.bfloat16 | (32, 16, 1, 1, 4096, 64)  |           37.671 |              40.928 |         0.920 |               0.749 |  13220.070 |
| noop          | torch.bfloat16 | (16, 16, 1, 1, 4096, 128) |           39.124 |              37.098 |         1.055 |               0.826 |  14528.175 |
| relative_bias | torch.bfloat16 | (16, 16, 1, 1, 4096, 128) |           39.093 |              37.235 |         1.050 |               0.823 |  14474.744 |
| head_bias     | torch.bfloat16 | (16, 16, 1, 1, 4096, 128) |           39.130 |              37.200 |         1.052 |               0.824 |  14488.547 |
| noop          | torch.bfloat16 | (8, 16, 1, 1, 16384, 64)  |           40.919 |              38.175 |         1.072 |               0.800 |  14173.236 |
| relative_bias | torch.bfloat16 | (8, 16, 1, 1, 16384, 64)  |           40.921 |              39.179 |         1.044 |               0.780 |  13810.127 |
| head_bias     | torch.bfloat16 | (8, 16, 1, 1, 16384, 64)  |           40.913 |              38.947 |         1.050 |               0.784 |  13892.216 |
| noop          | torch.bfloat16 | (4, 16, 1, 1, 16384, 128) |           43.949 |              37.232 |         1.180 |               0.820 |  14475.993 |
| relative_bias | torch.bfloat16 | (4, 16, 1, 1, 16384, 128) |           43.965 |              37.532 |         1.171 |               0.814 |  14360.240 |
| head_bias     | torch.bfloat16 | (4, 16, 1, 1, 16384, 128) |           43.967 |              37.214 |         1.181 |               0.821 |  14483.092 |
| noop          | torch.bfloat16 | (4, 16, 1, 1, 32768, 64)  |           44.986 |              38.249 |         1.176 |               0.798 |  14145.818 |
| relative_bias | torch.bfloat16 | (4, 16, 1, 1, 32768, 64)  |           44.999 |              38.620 |         1.165 |               0.791 |  14009.921 |
| head_bias     | torch.bfloat16 | (4, 16, 1, 1, 32768, 64)  |           44.999 |              39.345 |         1.144 |               0.776 |  13751.837 |
| noop          | torch.bfloat16 | (2, 16, 1, 1, 32768, 128) |           51.920 |              37.927 |         1.369 |               0.805 |  14210.515 |
| relative_bias | torch.bfloat16 | (2, 16, 1, 1, 32768, 128) |           51.930 |              38.026 |         1.366 |               0.803 |  14173.495 |
| head_bias     | torch.bfloat16 | (2, 16, 1, 1, 32768, 128) |           51.962 |              38.885 |         1.336 |               0.785 |  13860.444 |
| noop          | torch.float32 | (4, 16, 1, 16, 1024, 64) |          180.591 |              32.689 |         5.525 |               0.934 |    517.253 |
| relative_bias | torch.float32 | (4, 16, 1, 16, 1024, 64) |          180.632 |              32.802 |         5.507 |               0.931 |    515.463 |
| head_bias     | torch.float32 | (4, 16, 1, 16, 1024, 64) |          180.592 |              32.612 |         5.538 |               0.937 |    518.471 |
| noop          | torch.float32 | (1, 16, 1, 16, 4096, 64) |          680.394 |              36.482 |        18.650 |               0.837 |    463.474 |
| relative_bias | torch.float32 | (1, 16, 1, 16, 4096, 64) |          680.381 |              35.600 |        19.112 |               0.857 |    474.947 |
| head_bias     | torch.float32 | (1, 16, 1, 16, 4096, 64) |          680.446 |              35.502 |        19.167 |               0.860 |    476.266 |
| noop          | torch.float32 | (32, 16, 1, 2, 1024, 64) |          179.846 |              33.187 |         5.419 |               0.927 |   4075.827 |
| relative_bias | torch.float32 | (32, 16, 1, 2, 1024, 64) |          179.924 |              33.294 |         5.404 |               0.924 |   4062.753 |
| head_bias     | torch.float32 | (32, 16, 1, 2, 1024, 64) |          179.843 |              33.175 |         5.421 |               0.927 |   4077.417 |
| noop          | torch.float32 | (8, 16, 1, 2, 4096, 64)  |          677.202 |              36.871 |        18.367 |               0.829 |   3668.635 |
| relative_bias | torch.float32 | (8, 16, 1, 2, 4096, 64)  |          677.220 |              37.062 |        18.272 |               0.825 |   3649.686 |
| head_bias     | torch.float32 | (8, 16, 1, 2, 4096, 64)  |          677.194 |              35.845 |        18.892 |               0.853 |   3773.595 |
| noop          | torch.float32 | (2, 16, 1, 2, 16384, 64) |         2646.250 |              37.096 |        71.335 |               0.823 |   3646.395 |
| relative_bias | torch.float32 | (2, 16, 1, 2, 16384, 64) |         2645.856 |              37.189 |        71.146 |               0.821 |   3637.236 |
| head_bias     | torch.float32 | (2, 16, 1, 2, 16384, 64) |         2646.622 |              37.034 |        71.464 |               0.824 |   3652.469 |
| noop          | torch.float32 | (1, 16, 1, 2, 32768, 64) |         5366.302 |              36.646 |       146.435 |               0.833 |   3691.139 |
| relative_bias | torch.float32 | (1, 16, 1, 2, 32768, 64) |         5365.776 |              37.771 |       142.061 |               0.808 |   3581.235 |
| head_bias     | torch.float32 | (1, 16, 1, 2, 32768, 64) |         5366.462 |              37.660 |       142.498 |               0.811 |   3591.797 |
| noop          | torch.float32 | (64, 16, 1, 1, 1024, 64) |          179.938 |              35.190 |         5.113 |               0.881 |   7687.725 |
| relative_bias | torch.float32 | (64, 16, 1, 1, 1024, 64) |          179.855 |              33.394 |         5.386 |               0.928 |   8101.169 |
| head_bias     | torch.float32 | (64, 16, 1, 1, 1024, 64) |          179.991 |              35.271 |         5.103 |               0.879 |   7670.021 |
| noop          | torch.float32 | (16, 16, 1, 1, 4096, 64) |          676.176 |              37.952 |        17.817 |               0.807 |   7128.292 |
| relative_bias | torch.float32 | (16, 16, 1, 1, 4096, 64) |          676.166 |              37.189 |        18.182 |               0.824 |   7274.499 |
| head_bias     | torch.float32 | (16, 16, 1, 1, 4096, 64) |          676.218 |              38.025 |        17.783 |               0.806 |   7114.558 |
| noop          | torch.float32 | (4, 16, 1, 1, 16384, 64) |         2644.282 |              39.222 |        67.419 |               0.779 |   6897.511 |
| relative_bias | torch.float32 | (4, 16, 1, 1, 16384, 64) |         2642.099 |              37.351 |        70.738 |               0.818 |   7243.041 |
| head_bias     | torch.float32 | (4, 16, 1, 1, 16384, 64) |         2643.964 |              38.108 |        69.380 |               0.802 |   7099.051 |
| noop          | torch.float32 | (2, 16, 1, 1, 32768, 64) |         5358.217 |              39.348 |       136.174 |               0.776 |   6875.329 |
| relative_bias | torch.float32 | (2, 16, 1, 1, 32768, 64) |         5359.415 |              37.399 |       143.305 |               0.816 |   7233.773 |
| head_bias     | torch.float32 | (2, 16, 1, 1, 32768, 64) |         5359.189 |              39.134 |       136.945 |               0.780 |   6913.013 |
</details>

Chillee commented on benchmarks/transformer/score_mod.py:
```diff
@@ -175,6 +185,46 @@ def calculate_speedup(results: ExperimentResults, type: str) -> float:
+            * 2
+        )
+        output_size = query_size
+        total_size = (query_size + kv_size + output_size) / 1024 / 1024 / 1024  # In GB
```
GB is actually 1000^3, not 1024^3. GiB is 1024^3.

joydddd commented:
@pytorchbot merge

drisspg commented on benchmarks/transformer/score_mod.py:
```diff
@@ -62,6 +65,7 @@ class ExperimentResults:
 class Experiment:
     config: ExperimentConfig
     results: ExperimentResults
+    cal_bandwidth: bool
```
nit: this should probably go in experiment config

drisspg commented on benchmarks/transformer/score_mod.py:
```diff
@@ -175,6 +185,46 @@ def calculate_speedup(results: ExperimentResults, type: str) -> float:
+        raise ValueError(f"Invalid type {type}")
+
+
+def calculate_gflops(config: ExperimentConfig, results: ExperimentResults) -> float:
```
super nit: tflops might be more useful?

joydddd commented:
Sparsity Mask Performance:
(vs. sdpa) 
| Type    |   Speedup | score_mod   | dtype          | shape(B,Hq,M,Hkv,N,D)        |
|---------|-----------|-------------|----------------|------------------------------|
| Average |     1.320 |             |                |                              |
| Max     |     1.798 | causal_mask | torch.bfloat16 | (16, 16, 4096, 16, 4096, 64) |
| Min     |     0.928 | causal_mask | torch.bfloat16 | (2, 16, 512, 16, 512, 64)    |

<details> 
<summary> Full performance sweep </summary> 

| score_mod       | dtype          | shape(B,Hq,M,Hkv,N,D)         |   fwd_eager_time |   fwd_compiled_time |   fwd_speedup |
|-----------------|----------------|-------------------------------|------------------|---------------------|---------------|
| causal_mask     | torch.bfloat16 | (2, 16, 512, 16, 512, 64)     |           24.385 |              26.280 |         0.928 |
| causal_mask     | torch.bfloat16 | (2, 16, 512, 16, 512, 128)    |           35.409 |              31.516 |         1.124 |
| causal_mask     | torch.bfloat16 | (2, 16, 1024, 16, 1024, 64)   |           53.370 |              54.015 |         0.988 |
| causal_mask     | torch.bfloat16 | (2, 16, 1024, 16, 1024, 128)  |           87.119 |              88.522 |         0.984 |
| causal_mask     | torch.bfloat16 | (2, 16, 4096, 16, 4096, 64)   |          660.998 |             439.539 |         1.504 |
| causal_mask     | torch.bfloat16 | (2, 16, 4096, 16, 4096, 128)  |         1157.388 |             770.759 |         1.502 |
| causal_mask_gqa | torch.bfloat16 | (2, 16, 512, 2, 512, 64)      |           24.170 |              25.443 |         0.950 |
| causal_mask_gqa | torch.bfloat16 | (2, 16, 512, 2, 512, 128)     |           34.168 |              31.162 |         1.096 |
| causal_mask_gqa | torch.bfloat16 | (2, 16, 1024, 2, 1024, 64)    |           52.387 |              53.784 |         0.974 |
| causal_mask_gqa | torch.bfloat16 | (2, 16, 1024, 2, 1024, 128)   |           87.243 |              87.952 |         0.992 |
| causal_mask_gqa | torch.bfloat16 | (2, 16, 4096, 2, 4096, 64)    |          661.177 |             439.398 |         1.505 |
| causal_mask_gqa | torch.bfloat16 | (2, 16, 4096, 2, 4096, 128)   |         1157.591 |             770.911 |         1.502 |
| causal_mask     | torch.bfloat16 | (8, 16, 512, 16, 512, 64)     |           58.042 |              54.225 |         1.070 |
| causal_mask     | torch.bfloat16 | (8, 16, 512, 16, 512, 128)    |           93.491 |              88.991 |         1.051 |
| causal_mask     | torch.bfloat16 | (8, 16, 1024, 16, 1024, 64)   |          182.287 |             136.294 |         1.337 |
| causal_mask     | torch.bfloat16 | (8, 16, 1024, 16, 1024, 128)  |          310.577 |             238.393 |         1.303 |
| causal_mask     | torch.bfloat16 | (8, 16, 4096, 16, 4096, 64)   |         2576.397 |            1456.171 |         1.769 |
| causal_mask     | torch.bfloat16 | (8, 16, 4096, 16, 4096, 128)  |         4544.272 |            2751.014 |         1.652 |
| causal_mask_gqa | torch.bfloat16 | (8, 16, 512, 2, 512, 64)      |           55.248 |              53.256 |         1.037 |
| causal_mask_gqa | torch.bfloat16 | (8, 16, 512, 2, 512, 128)     |           91.206 |              87.683 |         1.040 |
| causal_mask_gqa | torch.bfloat16 | (8, 16, 1024, 2, 1024, 64)    |          180.697 |             135.871 |         1.330 |
| causal_mask_gqa | torch.bfloat16 | (8, 16, 1024, 2, 1024, 128)   |          309.582 |             238.091 |         1.300 |
| causal_mask_gqa | torch.bfloat16 | (8, 16, 4096, 2, 4096, 64)    |         2576.350 |            1458.435 |         1.767 |
| causal_mask_gqa | torch.bfloat16 | (8, 16, 4096, 2, 4096, 128)   |         4544.394 |            2751.855 |         1.651 |
| causal_mask     | torch.bfloat16 | (16, 16, 512, 16, 512, 64)    |          104.816 |              87.126 |         1.203 |
| causal_mask     | torch.bfloat16 | (16, 16, 512, 16, 512, 128)   |          171.095 |             151.639 |         1.128 |
| causal_mask     | torch.bfloat16 | (16, 16, 1024, 16, 1024, 64)  |          347.870 |             239.431 |         1.453 |
| causal_mask     | torch.bfloat16 | (16, 16, 1024, 16, 1024, 128) |          604.025 |             443.757 |         1.361 |
| causal_mask     | torch.bfloat16 | (16, 16, 4096, 16, 4096, 64)  |         5052.478 |            2809.603 |         1.798 |
| causal_mask     | torch.bfloat16 | (16, 16, 4096, 16, 4096, 128) |         8921.062 |            5391.688 |         1.655 |
| causal_mask_gqa | torch.bfloat16 | (16, 16, 512, 2, 512, 64)     |          101.131 |              86.103 |         1.175 |
| causal_mask_gqa | torch.bfloat16 | (16, 16, 512, 2, 512, 128)    |          170.009 |             150.509 |         1.130 |
| causal_mask_gqa | torch.bfloat16 | (16, 16, 1024, 2, 1024, 64)   |          346.097 |             238.924 |         1.449 |
| causal_mask_gqa | torch.bfloat16 | (16, 16, 1024, 2, 1024, 128)  |          603.510 |             443.726 |         1.360 |
| causal_mask_gqa | torch.bfloat16 | (16, 16, 4096, 2, 4096, 64)   |         5051.045 |            2812.355 |         1.796 |
| causal_mask_gqa | torch.bfloat16 | (16, 16, 4096, 2, 4096, 128)  |         8922.174 |            5393.416 |         1.654 |
</details>

joydddd committed:
Add flex decoding benchmark

ghstack-source-id: f33b5e489eb7e6b7baae306eb74de00c16251ddb
Pull Request resolved: https://github.com/pytorch/pytorch/pull/130788

joydddd committed:
Fix -nh arg

ghstack-source-id: d789dad142e946bf0c07af170baeea1909ab086b
Pull Request resolved: https://github.com/pytorch/pytorch/pull/130854

joydddd commented:
@pytorchbot merge

joydddd commented:
@pytorchbot merge
